src/app/components/RichText/index.tsx

```diff
@@ -0,0 +1,58 @@
+import { pipe, reduce } from 'ramda';
+import * as React from 'react';
+
+// TODO: generate from TL scheme
+export interface Entity {
+  _: string;
+  offset: number;
+  length: number;
+}
+
+interface Props {
+  entities: Entity[];
+  text: string;
+}
+
+type RichText = { _: string, text: string };
+type RichAccum = { list: RichText[], last: number };
+const parse = (text: string) => pipe(reduce<Entity, RichAccum, Entity[]>((acc, p) => {
+  return {
+    list: [...acc.list,
+      { _: 'messageEntityText', text: text.slice(acc.last, p.offset) },
+      {
+        _: p._,
+        text: text.slice(p.offset, p.offset + p.length),
+      },
+    ],
+    last: p.offset + p.length,
+  };
+}, { list: [], last: 0 }), p => p.list);
```

Comment:
Maybe it is worth making the parser logic itself a plain function? The reducer that will be created on each `parse` call will be much cheaper, and also, it will allow creating `formattedText` from props in one call

```typescript
type ParseReducer = (
  text  : string,
  list  : RichText[],
  last  : number,
  _     : string,
  offset: number,
  end   : number
) => RichAccum

const parseReducer: ParseReducer = (text, list, last, _, offset, end) => ({
  list: [
    ...list,
    richText('messageEntityText', text.slice(last, offset)),
    richText(_, text.slice(offset, end)),
  ],
  last: end,
})
const parse = (entities: Entity[], text: string) =>
  entities.reduce(
    ({ list, last }, { _, offset, length }) =>
      parseReducer(text, list, last, _, offset, offset + length),
    initial
  ).list

const richText = (_: string, text: string): RichText => ({ _, text })

const initial = { list: [], last: 0 }
```